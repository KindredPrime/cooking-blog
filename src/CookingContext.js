import React from 'react';

const CookingContext = React.createContext({
  user: null
});

export default CookingContext;