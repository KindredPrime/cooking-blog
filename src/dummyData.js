const dummyPosts = [
  {
    id: 1,
    lastEdited: 'Dec 17 07:37:16 2020',
    title: 'Lessons I Learned from Cooking',
    author: 'FakeUser2',
    content: `Have patience, it can be fun to experiment and explore, having a hard time learning something may only be temporary, bringing an idea of yours to life by creating it from scratch without relying on any instructions can be fun and rewarding, the meals you cook don't have to be fancy or difficult to taste good.`
  },
  {
    id: 2,
    lastEdited: 'Dec 28 18:24:43 2020',
    title: 'How to Save Money When Cooking',
    author: 'User1',
    content: `Buy in bulk, cook in bulk and freeze the leftovers, but store brand products.`
  },
  {
    id: 3,
    lastEdited: 'Dec 1 12:54:31 2020',
    title: '5 Easy Cooking Tips',
    author: 'User3',
    content: `1. When experimenting with seasonings in a recipe, mix them dry in a cup and smell it. If it smells good, then it'll probably taste good.
    2. Add some water when microwaving leftovers
    3. Start small with seasonings
    4. Cook
    5. Blah`
  },
  {
    id: 4,
    lastEdited: 'Jan 5 08:28:28 2021',
    title: 'Cooking Basics for Beginners',
    author: 'User1',
    content: `Go easy on seasonings and salt, if the inside of cooked chicken is pink or its juices aren't clear then it needs to be cooked longer, cooked ground beef needs to be brown in order to be safe to eat`
  },
  {
    id: 5,
    lastEdited: 'Dec 15 11:56:48 2020',
    title: 'Winter Seasoning Guide',
    author: 'User 2',
    content: `Rosemary, sage, thyme, cinnamon, nutmeg`
  },
  {
    id: 6,
    lastEdited: 'Jan 5 14:22:39 2021',
    title: 'Title6',
    author: 'User3',
    content: 'Post 6 content'
  },
  {
    id: 7,
    lastEdited: 'Jan 2 16:22:39 2021',
    title: 'Title7',
    author: 'User3',
    content: 'Post 7 content'
  },
  {
    id: 8,
    lastEdited: 'Jan 6 12:22:39 2021',
    title: 'Title8',
    author: 'User3',
    content: 'Post 8 content'
  },
  {
    id: 9,
    lastEdited: 'Dec 15 11:22:39 2021',
    title: 'Title9',
    author: 'User3',
    content: 'Post 9 content'
  },
  {
    id: 10,
    lastEdited: 'Dec 15 19:22:39 2021',
    title: 'Title10',
    author: 'User3',
    content: 'Post 10 content'
  },
  {
    id: 11,
    lastEdited: 'Jan 4 18:22:39 2021',
    title: 'Title11',
    author: 'User3',
    content: 'Post 11 content'
  }
];

const dummyComments = [
  {
    id: 1,
    lastEdited: new Date('Dec 19 07:37:16 2020'),
    content: 'These are all great lessons!',
    creator: 'User1',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 2,
    lastEdited: new Date('Dec 17 08:37:52 2020'),
    content: `I can't wait to try these seasoning combinations in my next meal!`,
    creator: 'User1',
    postTitle: 'Winter Seasoning Guide'
  },
  {
    id: 3,
    lastEdited: new Date('Jan 5 08:28:28 2021'),
    content: `I wish I knew these when I started learning to cook!`,
    creator: 'User4',
    postTitle: 'Cooking Basics for Beginners'
  },
  {
    id: 4,
    lastEdited: new Date('Dec 31 09:24:43 2020'),
    content: `Great advice!  This'll help me save so much money feeding my three teenagers!`,
    creator: 'User4',
    postTitle: `How to Save Money When Cooking`
  },
  {
    id: 5,
    lastEdited: new Date('Jan 1 10:31:39 2021'),
    content: `What an insightful post`,
    creator: 'FakeUser2',
    postTitle: 'Title7'
  },
  {
    id: 6,
    lastEdited: new Date('Dec 19 08:37:16 2020'),
    content: 'I know, right?',
    creator: 'User3',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 7,
    lastEdited: new Date('Dec 20 10:43:16 2020'),
    content: `It's amazing what you can learn from cooking`,
    creator: 'User4',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 8,
    lastEdited: new Date('Dec 17 20:12:16 2020'),
    content: 'Huh, I never thought about it this way',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 9,
    lastEdited: new Date('Dec 12 07:07:07 2020'),
    content: 'Comment9',
    creator: 'User6',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 10,
    lastEdited: new Date('Dec 19 15:22:16 2020'),
    content: 'Comment10',
    creator: 'User7',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 11,
    lastEdited: new Date('Dec 15 17:18:19 2020'),
    content: 'Comment11',
    creator: 'User8',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 12,
    lastEdited: new Date('Jan 1 12:13:14 2021'),
    content: 'Comment12',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 13,
    lastEdited: new Date('Jan 2 13:14:15 2021'),
    content: 'Comment13',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 14,
    lastEdited: new Date('Jan 5 06:07:08 2021'),
    content: 'Comment14',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 15,
    lastEdited: new Date('Jan 3 20:21:22 2021'),
    content: 'Comment15',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  }
];

const dummyUsers = [
  {
    id: 1,
    username: 'User1',
    password: 'password1',
    email: 'user1@gmail.com'
  },
  {
    id: 2,
    username: 'FakeUser2',
    password: 'password2',
    email: 'fakeuser2@gmail.com'
  },
  {
    id: 3,
    username: 'User3',
    password: 'password3',
    email: 'user3@gmail.com'
  },
  {
    id: 4,
    username: 'User4',
    password: 'password4',
    email: 'user4@gmail.com'
  },
  {
    id: 5,
    username: 'User5',
    password: 'password5',
    email: 'user5@gmail.com'
  },
  {
    id: 6,
    username: 'User6',
    password: 'password6',
    email: 'user6@gmail.com'
  },
  {
    id: 7,
    username: 'User7',
    password: 'password7',
    email: 'user7@gmail.com'
  },
  {
    id: 8,
    username: 'User8',
    password: 'password8',
    email: 'user8@gmail.com'
  }
];

export {
  dummyPosts,
  dummyComments,
  dummyUsers
};