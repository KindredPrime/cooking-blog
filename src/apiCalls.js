import { dummyPosts, dummyComments, dummyUsers } from './dummyData';

const blogPosts = dummyPosts;
const comments = dummyComments;
const users = dummyUsers;

let controller = new AbortController();
let signal = controller.signal;

function abortTasks() {
  controller.abort();
  controller = new AbortController();
  signal = controller.signal;
}

function getAllBlogPosts() {
  return new Promise((resolve, reject) => {
    signal.addEventListener('abort', () => {
      reject('getAllBlogPosts has been aborted');
    });

    resolve(blogPosts)
  });
}

function getBlogPostsByUser(username) {
  return new Promise((resolve, reject) => {
    const userPosts = blogPosts.filter((post) => post.author.username === username);

    signal.addEventListener('abort', () => {
      reject('getBlogPostsByUser has been aborted');
    });

    resolve(userPosts)
  });
}

function getBlogPostById(id) {
  return new Promise((resolve, reject) => {
    const blogPost = blogPosts.find((post) => post.id === parseInt(id));

    if (!blogPost) {
      reject(`There is no blog post with id ${id}`);
    }

    resolve(blogPost);
  });
}

function patchBlogPost(id, content) {
  return new Promise((resolve, reject) => {
    getBlogPostById(id)
      .then((blogPost) => {
        if (blogPost.content === content) {
          resolve(blogPost);
        }

        const newBlogPost = {
          ...blogPost,
          content,
          lastEdited: new Date()
        };

        // Update the post stored in this file
        const postsIndex = blogPosts.findIndex((elem) => elem.id === parseInt(id));
        blogPosts[postsIndex] = newBlogPost;

        resolve(newBlogPost);
      })
      .catch(reject);
  });
}

function getAllComments() {
  return Promise.resolve(comments);
}

function getCommentById(id) {
  return new Promise((resolve, reject) => {
    const comment = comments.find((comment) => comment.id === id);

    if (!comment) {
      reject(`Comment with id ${id} does not exist`);
    }

    resolve(comment);
  });
}

function getCommentsByBlogPost(title) {
  return new Promise((resolve, reject) => {
    const postComments = comments.filter((comment) => comment.blogPost.title === title);

    resolve(postComments);
  });
}

function getCommentsByUser(username) {
  return Promise.resolve(comments.filter((comment) => comment.creator.username === username));
}

function addComment(comment) {
  return new Promise((resolve, reject) => {
    const { content, creator, blogPost } = comment;
    const id = comments.length + 1;

    const newComment = {
      id,
      lastEdited: new Date(),
      content,
      creator,
      blogPost
    }

    comments.push(newComment);

    resolve(newComment);
  });
}

function patchComment(id, content) {
  return new Promise((resolve, reject) => {
    getCommentById(id)
      .then((comment) => {
        if (comment.content === content) {
          resolve(comment);
        }

        const newComment = {
          ...comment,
          content,
          lastEdited: new Date()
        };

        // Update the comment stored in this file
        const commentsIndex = comments.findIndex((elem) => elem.id === id);
        comments[commentsIndex] = newComment;

        resolve(newComment);
      })
      .catch(reject);
  });
}

function deleteComment(id) {
  return new Promise((resolve, reject) => {
    const commentsIndex = comments.findIndex((comment) => comment.id === id);

    if (commentsIndex < 0) {
      reject(`Comment with id ${id} does not exist`);
    }

    const comment = comments[commentsIndex];
    comments[commentsIndex] = {
      ...comment,
      lastEdited: new Date(),
      content: '[Deleted]',
      creator: null,
      deleted: true
    };

    resolve();
  });
}

function getUserById(id) {
  return Promise.resolve(users.find((user) => user.id === parseInt(id)));
}

export {
  abortTasks,
  getAllBlogPosts,
  getBlogPostsByUser,
  getBlogPostById,
  patchBlogPost,
  getAllComments,
  getCommentById,
  getCommentsByBlogPost,
  getCommentsByUser,
  addComment,
  patchComment,
  deleteComment,
  getUserById
};