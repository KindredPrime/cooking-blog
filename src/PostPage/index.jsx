import { Component } from 'react';
import PropTypes from 'prop-types';
import CookingContext from '../CookingContext';
import * as API from '../apiCalls';
import EditPost from '../EditPost/index';
import CommentsList from '../CommentsList/index';
import './index.css';

class PostPage extends Component {
  static contextType = CookingContext;

  state = {
    isEditingPost: false,
    lastEdited: null,
    title: null,
    author: null,
    content: null,
    comments: []
  };

  handleEditCancel = () => {
    this.setState({
      isEditingPost: false
    });
  }

  handleEditSubmit = (content) => {
    const { id } = this.props.match.params;

    API.patchPostById(id, content)
      .then(() => {
        this.setState({
          content,
          isEditingPost: false
        });
      })
      .catch(console.log);
  }

  componentDidMount() {
    const { id } = this.props.match.params;

    API.getPostById(id)
      .then((post) => {
        const { lastEdited, title, author, content } = post;

        this.setState({
          lastEdited,
          title,
          author,
          content
        });

        return title;
      })
      .then((title) => API.getCommentsByPost(title))
      .then((comments) => this.setState({ comments }))
      .catch((error) => console.log(error));
  }

  componentWillUnmount() {
    API.abortTasks();
  }

  render() {
    const { isEditingPost, lastEdited, title, author, content, comments } = this.state;
    const { username } = this.context;

    return (
      <main className="PostPage">
        <h1>{title}</h1>
        <h2>By {author}</h2>

        <p>Last edited on {lastEdited}</p>

        {isEditingPost
          ? <EditPost content={content} handleCancel={this.handleEditCancel} handleSubmit={this.handleEditSubmit} />
          : <>
            <p className="PostPage__content">{content}</p>

            {/* Display an edit button if the current user is the author */}
            {(username && username === author) &&
              <button
                type="button"
                onClick={() => this.setState({ isEditingPost: true })}
              >
                Edit
              </button>}
          </>}

        <CommentsList comments={comments} />
      </main>
    );
  }
}

PostPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    })
  })
};

export default PostPage;