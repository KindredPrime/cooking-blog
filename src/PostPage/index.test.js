import ReactDOM from 'react-dom';
import { render, waitFor } from '@testing-library/react';
import * as API from '../apiCalls';
import { dummyPosts } from '../dummyData';
import PostPage from './index';
import CookingContext from '../CookingContext';

describe('PostPage Component', () => {
  const origAPI = API;

  /**
   * Rewrite API methods to mock their results
   */
  beforeAll(() => {
    API.getPostById = (id) => new Promise((resolve, reject) => {
      resolve(dummyPosts.find((post) => post.id === parseInt(id)));
    });
  });

  /**
   * Reset API methods to their original values
   */
  afterAll(() => {
    API.getAllPosts = origAPI.getAllPosts;
    API.getPostsByUser = origAPI.getPostsByUser;
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <PostPage match={{ params: { id: '1' } }} />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it(`renders the post as expected, where the user is NOT the post's author`, async () => {
    const contextValue = {
      username: 'not-the-author'
    };
    const id = 1;

    render(
      <CookingContext.Provider value={contextValue}>
        <PostPage match={{ params: { id } }} />
      </CookingContext.Provider>
    );

    await waitFor(() => expect(dummyPosts[id-1].title).toBeInTheDocument);
    expect(document.body).toMatchSnapshot();
  });

  it(`renders the post as expected, where the user IS the post's author`, async () => {
    const id = 1;
    const contextValue = {
      username: dummyPosts[id-1].author
    };

    render(
      <CookingContext.Provider value={contextValue}>
        <PostPage match={{ params: { id } }} />
      </CookingContext.Provider>
    );

    await waitFor(() => expect(dummyPosts[id-1].title).toBeInTheDocument);
    expect(document.body).toMatchSnapshot();
  });
});