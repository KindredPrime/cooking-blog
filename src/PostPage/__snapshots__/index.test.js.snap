// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostPage Component renders the post as expected, where the user IS the post's author 1`] = `
<body>
  <div>
    <main
      class="PostPage"
    >
      <h1>
        Lessons I Learned from Cooking
      </h1>
      <h2>
        By 
        FakeUser2
      </h2>
      <p>
        Last edited on 
        Dec 17 07:37:16 2020
      </p>
      <p
        class="PostPage__content"
      >
        Have patience, it can be fun to experiment and explore, having a hard time learning something may only be temporary, bringing an idea of yours to life by creating it from scratch without relying on any instructions can be fun and rewarding, the meals you cook don't have to be fancy or difficult to taste good.
      </p>
      <button
        type="button"
      >
        Edit
      </button>
      <div
        class="CommentsList"
      />
    </main>
  </div>
</body>
`;

exports[`PostPage Component renders the post as expected, where the user is NOT the post's author 1`] = `
<body>
  <div>
    <main
      class="PostPage"
    >
      <h1>
        Lessons I Learned from Cooking
      </h1>
      <h2>
        By 
        FakeUser2
      </h2>
      <p>
        Last edited on 
        Dec 17 07:37:16 2020
      </p>
      <p
        class="PostPage__content"
      >
        Have patience, it can be fun to experiment and explore, having a hard time learning something may only be temporary, bringing an idea of yours to life by creating it from scratch without relying on any instructions can be fun and rewarding, the meals you cook don't have to be fancy or difficult to taste good.
      </p>
      <div
        class="CommentsList"
      />
    </main>
  </div>
</body>
`;
